operation/manager.go

```diff
@@ -41,32 +43,42 @@ const somCtxKeySingleOp = somCtxKey(iota)
 
 // New creates a new operation, cancels previous, returns a new context and function to call when done.
 func (sm *SingleOperationManager) New(ctx context.Context) (context.Context, func()) {
-	// handle nested ops
+	// Handle nested ops. Note an operation set on a context by one `SingleOperationManager` can be
+	// observed on a different instance of a `SingleOperationManager`.
 	if ctx.Value(somCtxKeySingleOp) != nil {
 		return ctx, func() {}
 	}
 
 	sm.mu.Lock()
 
-	// first cancel any old operation
-	sm.cancelInLock(ctx)
+	// Cancel any existing operation. This blocks until the operation is completed.
+	if sm.currentOp != nil {
+		sm.currentOp.cancelFunc()
+	}
 
-	theOp := &anOp{}
+	theOp := &anOp{
+		closedCond: sync.NewCond(&sm.mu),
```

Comment:
I preferred to add the `closedCond` variable onto the `SingleOperationManager` structure. Which would avoid an allocation per operation.

However, condition variables need to have their constructor called. And currently `SingleOperationManager` is default constructible. So having the condition variable on the `SingleOperationManager` would require finding and changing all creations of a `SingleOperationManager`.

Reply:
Would that be constrained to in rdk constructors (if you mean resources that are using that interface) or would it be a complete API change that affects modules as well?

Reply:
I'd actually just prefer we make the allocation in the form of a `chan struct{}` that gets closed on the cancel func on the new end and received upon on the theOp.cancelFunc side

Reply:
> Would that be constrained to in rdk constructors (if you mean resources that are using that interface) or would it be a complete API change that affects modules as well?

I'm not 100% sure I've read the question correctly ("what would be constrained?"). But I think* the answer is that none of the code paths into module APIs use a `SingleOperationManager` today. We could hypothetically wrap those code paths with one (for example, a dedicated SingleOpManager per configured module). But that was a non-goal for the project.

> I'd actually just prefer we make the allocation in the form of a chan struct{} that gets closed on the cancel func on the new end and received upon on the theOp.cancelFunc side

I'm double checking this, but it's not as attractive if we need to keep the SingleOpManager's mutex around. If we need to keep the mutex around, we need to be careful that we unlock/relock in the spots around the waiters receiving on the channel. Failure to do so is a recipe for deadlocks (holding resources while performing blocking operations -- you can thank Geert for beating me over the head with this). 

Condition variables take care of the locking on your behalf.

Reply:
Just double checked. I think this option of keeping the condition variable on the operation is a bug (specifically: that in conjunction with considering the "waiting" done only when the [prior op completes](https://github.com/viamrobotics/rdk/pull/2667/files#diff-6163f94661630679b64f82eba6259cc10b6cbec32a16fa47326e87a5f61d6b8cR70).)

The channel doesn't fix that. I'll try to push something out today so you can take a look, but in lieu of that we can talk offline about condition variable/channel.

Reply:
> If we need to keep the mutex around
is there a scenario where we don't want to keep it around?

Reply:
> is there a scenario where we don't want to keep it around?

The `SingleOpManager.currentOp` is the only shared variable here (everything is derivative from that value). The mutex could be removed and instead have the code access `currentOp` with [compare and swap](https://pkg.go.dev/sync/atomic#CompareAndSwapPointer).

I don't think that solution is appropriate here. It's a more advanced technique and we'd still need to write some sort of "wait for the `currentOp` value to be nil" -- either as a spin loop or a combination of mutexes/cond variables/go channels. So it just moves the complexity around.

That said, I'm happy to sketch the change offline for anyone that wants an appreciation for the code alternatives and how they look.

Reply:
cool, the optimization sounds like something we should do

Reply:
To reduce the risk of moving the condition variable onto the `SingleOpManager`, I've created https://viam.atlassian.net/browse/RSDK-4486 which I'll take care of right after this is pushed.